Add initial sphere-voxel collision detection

patowen opened the pull request:
This PR adds collision detection logic. There is a `shape_cast` method in the `common::collision` module that answers the question of "How far can a sphere of a given radius move in a given direction, up to a given distance, before colliding with a voxel? Does it collide at all?" If there is a collision, it also finds the normal, which can be used for collision resolution.

For testing purposes, a basic form of collision resolution is also included with these changes. As long as noclip is disabled by pressing "V", if a character collides with a wall, it will stop. This has the advantage of simplicity, but it has the disadvantage of making walls "sticky", so smoother movement is planned in a future PR.

Ralith commented on common/src/ray_tracing.rs:
```diff
@@ -0,0 +1,105 @@
+/// normals are given in. Specifically, the `start_node_transform` matrix converts this coordinate system to node coordinates.
+///
+/// The `tanh_distance` is the hyperbolic tangent of the distance along the ray to check for intersections.
+pub fn trace_ray(
```
This function *demands* a decent suite of unit tests.

patowen replied:
The logic that definitely needs unit tests in this function gets moved to `sphere_collider.rs` later on. My plan for now is to create a unit test in this commit and move/adapt it in the future commit. It's possible we'll still want this method to be tested directly in the final result, but I'll hold off on that for now.

Ralith replied:
Playtesting is an extremely inefficient way to test. We should cover at least basic sanity checks, and ideally interesting corner cases as well.

patowen replied:
I'm happy to create a few basic sanity checks. However, the fact that we want test suites on functions that act on a populated `DualGraph` suggests that a future enhancement outside the scope of this PR would be the ability to select a "test world" for worldgen, and this can be the same world that is used for these kinds of tests.

For now, I can create a basic test world with only two or three solid blocks, although this will still be a fair amount of setup code.

Ralith replied:
Specifying geometry to test against is certainly necessary. We should work towards factoring out common code for that as test cases grow.

patowen replied:
I managed to add a test case for this method which tests a fair amount of its functionality. To save my own time, my plan is to wait until you've reviewed some of my tests before I add some for corner cases. That way, I can get a better sense of your preferred coding style when it comes to tests for this project.

patowen replied:
Based on your review, it doesn't seem like there's anything fundamentally wrong with how I set up these tests, so I'll add some edge case tests here before I request another review.

patowen replied:
I found a bug while testing an edge case, so you were definitely right that this needed testing. When setting `ray_node_distance`, I used the wrong formula for "distance to the origin". For a normalized hypervector, it should either by `.xyz().magnitude().asinh()` or `.w.cosh()`, but I used `.xyz().magnitude().acosh()`.

I tried to test for this manually, but I guess my aim wasn't good enough. Thanks for insisting on unit tests!

Ralith commented on common/src/character_controller.rs:
```diff
@@ -77,4 +89,66 @@ impl CharacterControllerPass<'_> {
+struct Collision {
+    /// This collision normal faces away from the collision surface and is given in the perspective of the character
+    /// _after_ it is transformed by `allowed_displacement`.
+    normal: na::Vector3<f32>,
```
Why is this a `Vector3` when `RayTracingIntersection`'s is a `Vector4`?

patowen replied:
The normal vector here is in relation to the character, and the implication is that the vector represents a direction starting at the character's center. Because of that, the `w` coordinate is always 0, so we enforce that invariant by making it a `Vector3` instead of a `Vector4`. Converting a `Vector4` to this `Vector3` effectively performs the necessary projection to ensure this invariant holds.

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,294 @@
+use crate::{
```
Approximately everything in this file needs unit tests.

Ralith commented on common/src/ray_tracing.rs:
```diff
@@ -103,3 +178,82 @@ impl std::ops::Mul<&Ray> for na::Matrix4<f32> {
 }
+
+/// Represents a discretized region in the voxel grid contained by an axis-aligned bounding box.
+pub struct CubicVoxelRegion {
```
`VoxelAABB`?

Ralith commented on common/src/ray_tracing.rs:
```diff
@@ -19,39 +32,81 @@ pub fn trace_ray(
+/// Contains all the immutable data needed for `ChunkRayTracer` to perform its logic
+pub struct ChunkRayTracingContext<'a> {
+    pub dimension: usize,
+    pub dimension_f32: f32,
```
Storing these separately is not necessary. Seems like mostly code wants the `f32`, so let's stick with that.

patowen replied:
If we want to keep just one of them, I think it makes more sense to keep `dimension`, since it is used in some places, and converting a usize to a float and back to a usize without any other changes seems not ideal.

This context class isn't permanently stored anywhere, so there doesn't seem to be much of a downside to keeping both versions of `dimension`.

That being said, I'm also happy to just use `dimension as f32` everywhere.

Ralith replied:
Sure, sounds good.

> there doesn't seem to be much of a downside to keeping both versions of dimension.

The downside is increased verbosity, decreased readability, and room for error.

Ralith reviewed (changes requested):
Initial incomplete pass; more to come over coming days, but this should get things rolling.

Ralith commented on common/src/ray_tracing.rs:
```diff
@@ -1,13 +1,26 @@
-/// direction to detect an intersection.
+/// Performs ray tracing against the voxels in the `DualGraph`. This function is suitable for collision checking with
+/// any different collider shape. Each collider would have a different implemenation of `ChunkRayTracer`, which describes
+/// how the collision checking math works within a single chunk.
```
Do we have immediate plans for any other shapes to be tested against voxels? If not, a trait seems very premature.

patowen replied:
We will want to test points for the ray-casting used to allow block placing and breaking.

patowen replied:
The use case already exists in the collision-demo branch. Any logic involving rays can benefit from the graph traversal algorithm in ray_tracing.rs.

However, I understand that the design can change far more than originally anticipated, so I'm okay with getting rid of this abstraction.

Ralith replied:
It sounds like we'll eventually want to factor out the graph traversal algorithm into a more abstract swept broadphase, then. I'd definitely favor leaving the abstraction for that until we actually need it.

Ralith replied:
"swept broadphase" is a terse description of the graph traversal algorithm.

Ralith commented on common/src/ray_tracing.rs:
```diff
@@ -1,13 +1,26 @@
-/// Performs ray tracing against the z=0 plane in node coordinates. This plane is one-sided, so the ray must be going in the -z
-/// direction to detect an intersection.
+/// Performs ray tracing against the voxels in the `DualGraph`. This function is suitable for collision checking with
+/// any different collider shape. Each collider would have a different implemenation of `ChunkRayTracer`, which describes
```
Is this a ray trace or a shape cast (aka swept collision query)? The name suggests the former, but this comment, and the presence of collision shape information other than the voxels, suggests the latter.

patowen replied:
This is a shape cast, assuming the collider used isn't a point collider. I am not familiar with the terms shape-casting and "swept collision query", so ray tracing was the best term I could come up with.

I've been visualizing it as ray tracing against an abstract shape where a constant-width outline is added around all the voxels.

Ralith replied:
Let's call it "shape cast" then.

patowen commented on common/src/ray_tracing.rs:
```diff
@@ -0,0 +1,105 @@
+}
+
+/// Contains all information that was discovered as a result of ray tracing.
+pub struct RayTracingResult {
```
I thought of `RaySegment` but decided against it because it implied that it contained both endpoints. I like the name you suggested.

patowen commented on common/src/ray_tracing.rs:
```diff
@@ -0,0 +1,105 @@
+}
+
+impl Ray {
+    pub fn new(position: na::Vector4<f32>, direction: na::Vector4<f32>) -> Ray {
```
Possibly, although I don't think we `debug_assert` most normalization invariants in the rest of the codebase.

Ralith replied:
Might not be a bad time to start.

patowen replied:
I'm having second thoughts about implementing this.
* Approximate equality is needed for this debug assertion.
* `approx` is currently a dev-dependency, so it only works in tests.
* `approx` has no `debug_assert_abs_diff_eq!` method, so I would need to use `debug_assert!(abs_diff_eq!(...))` instead, making error messages less useful unless additional boilerplate code is added.
* I currently highly doubt this will be a common source of bugs. Trying to make this more airtight might be better done when we start having separate types for various normalization states as you commented below.

If you would still prefer me to add this, which approach would you recommend. Should I add the dependency, add new boilerplate code somewhere (where would you recommend?), or just do something like the following?
```rust
debug_assert!(
    (math::mip(&position, &position) + 1.0).abs() < 1e-5,
    "mip(position, position) == {}",
    math::mip(&position, &position)
);
```

Ralith replied:
I had imagined we'd do something like your example, but I agree it's low risk and redundant to more rigorous mechanisms involving the type system.

patowen commented on common/src/ray_tracing.rs:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    /// Returns a point along this ray `atanh(tanh_distance)` units away from the origin. This point
+    /// is _not_ lorentz normalized.
```
Maybe? One thing in that vein that I had considered is that we could have a type for a "vector with Lorentzian semantics", which would wrap an `na::Vector4<N>` and have member functions for operations that make sense in hyperbolic geometry, like `mip`. Then, the "Unit" version of this type would be Lorentz-normalized. There could even be separate types for point-like unit vectors (mip with itself is `-1`) and direction-like vectors (mip with itself is `+1`).

Ralith replied:
I like the sound of that. Something for hyperbolic isometries, too. Later, though.

patowen replied:
Created https://github.com/Ralith/hypermine/issues/264.

patowen commented on common/src/ray_tracing.rs:
```diff
@@ -0,0 +1,105 @@
+use crate::math;
+
+/// Performs ray tracing against the z=0 plane in node coordinates. This plane is one-sided, so the ray must be going in the -z
+/// direction to detect an intersection.
```
It would be the -z direction based on the Hyperboloid model. This would also be the direction represented by the `(0, 0, -1, 0)` vector, which is a valid direction vector, as it's tangent to the hyperboloid whenever we're on the z=0 plane.

patowen commented on common/src/ray_tracing.rs:
```diff
@@ -0,0 +1,105 @@
+/// direction to detect an intersection.
+///
+/// The `start_node_transform` parameter determines which coordinate system the `ray` parameter and any resulting intersection
+/// normals are given in. Specifically, the `start_node_transform` matrix converts this coordinate system to node coordinates.
```
If we separated these concerns, I don't think it would save that much work. The reason why `ray_tracing` handles these transformations might not immediately be clear until the graph traversal commit, when there are multiple `node_transform`s.

Ralith replied:
Checks out.

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,294 @@
+/// Finds the tanh of the distance a sphere will have to travel along a ray before it
+/// intersects the given point.
+///
+/// Returns NaN if there's no such intersection
```
It's very easy to handle possibly-`NaN` data incorrectly. Should this and below to return `Option<f32>` instead?

patowen replied:
Possibly. It would make the code calling it slightly easier to read, and any performance cost should be minor compared to the vector arithmetic required.

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,294 @@
+/// Converts from t-u-v coordinates to x-y-z coordinates. t-u-v coordinates are a permuted version of x-y-z coordinates.
+/// `t_axis` determines which of the three x-y-z coordinates corresponds to the t-coordinate. This function works with
+/// any indexable entity with at least three entries. Any entry after the third entry is ignored.
+fn tuv_to_xyz<T: std::ops::IndexMut<usize, Output = N>, N: Copy>(t_axis: usize, tuv: T) -> T {
```
`fn tuv_to_xyz<T: Copy, const D: usize>(t_axis: usize, tuv: [T; D])`, then calling `.into()` on vectors where needed, might be a little more straightforward. Your call, though.

Ralith commented on common/src/ray_tracing.rs:
```diff
@@ -103,3 +178,82 @@ impl std::ops::Mul<&Ray> for na::Matrix4<f32> {
+    }
+
+    /// Creates an iterator over grid points contained in the region, represented as ordered triples
+    pub fn grid_point_iterator(
```
nit: `grid_points` is more concise and no less clear; similarly below.

Ralith commented on common/src/ray_tracing.rs:
```diff
@@ -24,48 +24,118 @@ pub fn trace_ray(
+                    // Ray cannot intersect node
+                    continue;
+                }
+                // If we have to do collision checking on nodes that don't exist in the graph, we cannot have a conclusive result.
```
Could we treat missing nodes/unpopulated chunks as solid walls for collision detection purposes?

patowen replied:
It's definitely possible, but it would be a compromise. Server-side, it's always a logic error if we collide with an unpopulated chunk, since we generate synchronously. This may be worth discussing pros and cons on Discord.

patowen commented:
I've updated this PR and added most of the necessary tests. The main shape casting method in the collision module still needs test, as I've saved that for last, and it might take some time. I believe there are enough new changes that it's worth a review (especially reviewing the design of the test). When I find time, I'll add the remaining missing tests.

I believe everything other than `collision::shape_cast` is well-covered.

Ralith commented on client/src/graphics/voxels/mod.rs:
```diff
@@ -146,27 +146,24 @@ impl Voxels {
                             if self.worldgen.load(ChunkDesc { node, params }).is_ok() {
-                                sim.graph.get_mut(node).as_mut().unwrap().chunks[chunk] =
-                                    Generating;
+                                *sim.graph.get_chunk_mut(chunk).unwrap() = Generating;
```
nit: Perhaps we should `impl Index(Mut)<ChunkId>` for panicing assets to save writing `unwrap` a bunch.

patowen replied:
That seems like a good idea. There is the risk of reduced mindfulness when making `unwrap` implicit. However, this tradeoff has already been considered in Rust standard library, and they decided in favor of allowing this, so I am willing to believe the benefits outweigh the costs.

Ralith commented on common/src/collision.rs:
```diff
@@ -0,0 +1,129 @@
+
+/// Where a cast ray ended, and all information about the hit at the end if such a hit occurred.
+#[derive(Debug)]
+pub struct RayEndpoint {
```
s/`Ray`/`Cast`/ for consistency

patowen replied:
I had left it as `Ray` for consistency with the `Ray` struct, since this class can act as a companion to it. Still happy to rename, but I would like your confirmation.

Ralith replied:
I think a more generic name makes sense. You can cast rays and spheres and other things, and it's weird to use ray-oriented nomenclature to refer to a sphere-cast result which isn't, in fact, on the ray.

patowen replied:
That's fair. I'll use the term `Cast` for everything except the `Ray` struct.

Ralith commented on common/src/collision.rs:
```diff
@@ -0,0 +1,129 @@
+
+/// Information about the intersection at the end of a ray segment.
+#[derive(Debug)]
+pub struct RayHit {
```
as above

Ralith commented on common/src/collision.rs:
```diff
@@ -0,0 +1,129 @@
+}
+
+#[cfg(test)]
+mod test {
```
This is a good start, but we should always also test *negative* cases. For example, a ray parallel to the plane (though I guess that's less interesting in H3, heh) and a ray pointing away from it.

patowen replied:
Oh yeah. I didn't try that hard to add extensive tests here because the code was going to be deleted.

Ralith commented on common/src/collision.rs:
```diff
@@ -1,14 +1,25 @@
 /// The `tanh_distance` is the hyperbolic tangent of the distance along the ray to check for hits.
+///
+/// This implementaion is incomplete, and it will only detect intersection with voxels in `start_chunk`.
 pub fn shape_cast(
```
nit: Since this is no longer generic, `sphere_cast` would be a bit clearer.

Ralith commented on common/src/collision.rs:
```diff
@@ -109,21 +178,140 @@ impl std::ops::Mul<&Ray> for na::Matrix4<f32> {
+        )
+        .unwrap();
+
+        let covered_planes: HashSet<_> = aabb.grid_planes(0).collect();
```
`collect`ing this into a `HashSet` feels a bit silly when it starts out as a `Range` which is both smaller and faster to query against. Perhaps returning `impl Iterator` is excessively opaque in this case?

patowen replied:
I'm undecided on this. The advantage of `impl Iterator` is that it's consistent with the other `VoxelAABB` functions.

The speed disadvantage is minor, since it's still very fast, and tests need not be optimized unless they're slowing down CI.

If I test with lines or points, I would need a hashset.

Ralith replied:
No sense overthinking support code in a test I suppose.

Ralith commented on common/src/collision.rs:
```diff
@@ -109,21 +178,140 @@ impl std::ops::Mul<&Ray> for na::Matrix4<f32> {
+        )
+        .unwrap();
+
+        let covered_planes: HashSet<_> = aabb.grid_planes(0).collect();
```
Should we test the other axes too?

patowen replied:
We can. I consider all axes equivalent with their logic, so it's not entirely necessary. However, we should probably test with a few different rays, and we might as well have those tests look at different axes.

Ralith replied:
It's not obvious to me that it'd be impossible to have bugs that break some, but not all, axes, and it seems easy enough to just slap a for loop around this whole thing.

patowen replied:
Good point. Since the ray isn't equal in all coordinates, it would give us more test cases for free.

patowen replied:
Actually, looking at this again, the test for planes isn't quite complete, as an implementation that passes this test for all rays could still be buggy. The planar bounds of a sphere's path is not determined just by the start and the end point, as cylinders bulge in the middle. This assumption only works in Euclidean geometry.

Rather than try some difficult-to-verify calculation to see whether a plane is in the sphere's path, I think the right move is to just test the union of a bunch of spheres along this path, since then we just need sphere/plane intersection queries. This would also let me test lines and points without hitting the wall I was hitting earlier.

Ralith replied:
Sounds reasonable to me.

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,800 @@
+};
+
+/// Handles collisions for a single chunk.
+pub fn chunk_shape_cast(ctx: &ChunkShapeCastingContext, endpoint: &mut RayEndpoint) {
```
nit: `sphere_cast`

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,800 @@
+    // We use an alternative quadratic formula to ensure that we return a positive number if `constant_term > 0.0`.
+    // Otherwise, the edge case of a small positive `constant_term` could be mishandled.
+    // Note that discriminant can be negative, which allows this function to return NaN when there is no solution.
+    Some(constant_term / (-half_linear_term + discriminant.sqrt()))
```
Can both `half_linear_term` and `discriminant` be zero here?

Ralith replied:
Also, could `half_linear_term` be equal to `discriminant.sqrt()`?

patowen replied:
To get an infinity from this, `half_linear_term` has to be non-negative. This would only happen if the character is moving away from a wall, and no collision should be detected.

Infinity is outside the bounds of any range we choose, so this edge case shouldn't give us problems.

Ralith replied:
I suppose infinity is a relatively sensible/harmless result.

patowen replied:
I definitely appreciate the call-out. Anywhere infinities can get involved, you can also end up with large numbers where you didn't intend them. I did need to think about it to be sure that we were fine after reading your question.

I think this is probably worth a code comment.

patowen replied:
Actually, thinking about this again, I think this does need to be addressed. If the numerator is also zero, we get NaN, and we cannot necessarily rule that out. I'll take a closer look at this.

patowen replied:
It looks like the answer is to restrict the solution space to non-negative lower solutions that are not double-roots (We don't care about double-roots because they correspond to moving tangent to a surface, which doesn't have to be a collision). This forces the constant term to be non-negative and the linear term to be negative. Because the linear term is negative, the denominator cannot be zero.

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,800 @@
+                transform,
+            };
+
+            // Populate voxels. Consists of a single cube with grid coordinates from (1, 1, 1) to (2, 2, 2)
```
What space are the coordinates in this comment? The context suggests voxel coordinates, but only a single voxel is set, not a box of eight.

patowen replied:
They're in "grid coordinates". the voxel coordinates would just be (2, 2, 2).

Ralith replied:
Is that scheme used anywhere else?

patowen replied:
It was one of the coordinate systems discussed in Discord, and it's used for the iterators in `VoxelAABB`.

Ralith replied:
Let's revisit this in the future I guess. Probably mostly a doc problem.

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,800 @@
+
+    /// Tests that a suitable collision is found when approaching a single voxel from various angles.
+    #[test]
+    fn chunk_shape_cast_examples() {
```
A few negative cases here wouldn't go amiss.

patowen replied:
I do have negative cases in separate tests, but I can add one here as well.

Ralith commented on common/src/collision.rs:
```diff
@@ -23,54 +23,122 @@ pub fn shape_cast(
+    // Start a breadth-first search of the graph's chunks, performing collision checks in each relevant chunk.
+    // The `chunk_queue` contains ordered pairs containing the `ChunkId` and the transformation needed to switch
+    // from the original node coordinates to the current chunk's node coordinates.
+    let mut visited_chunks: HashSet<ChunkId> = HashSet::new();
```
nit: `FxHashSet` is a better choice when the input isn't at risk of being carefully crafted to produce a DoS.

patowen replied:
I originally chose `HashSet` because this is a test. I'll switch it over.

Ralith replied:
`pub fn shape_cast` is a test?

patowen replied:
Oh, whoops. I got this mixed up with the `VoxelAABB` test.

Ralith reviewed (changes requested):
This is looking great, thanks!

Ralith commented on common/src/sphere_collider.rs:
```diff
@@ -0,0 +1,800 @@
+
+    // We use an alternative quadratic formula to ensure that we return a positive number if `constant_term > 0.0`.
+    // Otherwise, the edge case of a small positive `constant_term` could be mishandled.
+    // Note that discriminant can be negative, which allows this function to return NaN when there is no solution.
```
Outdated comment.

patowen commented:
Another round of changes has been complete. This time, I have added all the tests that I believe should be added, but I am open to suggestions if you think I missed something important. I also believe all your other suggestions have been addressed.

I made a few other minor changes as well that I noticed while rereading the code. Probably the most important change is that `sphere_cast` now takes a Position `position` instead of a matrix `start_node_transform` and a ChunkId `start_chunk`. There are at least two benefits to this: (1) The caller doesn't have to awkwardly pass in an arbitrary vertex, as that is handled internally, and (2) It's much easier to explain what the `position` parameter means in the documentation.

patowen commented:
@Ralith, looking at this code again, I think it might make more sense to try to make sphere_collider.rs more stand-alone. It doesn't need a `ChunkId` or a transformation matrix to compute anything internally; that's only used on the result. Is this something worth pursuing in this PR, or would you prefer it as future refactor in a separate PR?

Ralith commented:
That sounds like a nice improvement. I don't think it needs to block this PR, but if you get it into a state you're happy with before we merge this, then you might as well fold them together.

patowen commented:
> That sounds like a nice improvement. I don't think it needs to block this PR, but if you get it into a state you're happy with before we merge this, then you might as well fold them together.

Sounds good to me. Once you're reviewed it to your satisfaction and believe it is ready to merge as-is, I think the right move is to merge it, with the spirit of not letting "perfect" be the enemy of "good".

However, if I find time to make this adjustment before we reach that point, I'll go ahead and add it to this PR, adjusting the commit history as if this was the plan from the start. If we go this route, it may be a good idea for me to prepare another diff for you to reference so that it's easier for you to see what changed.

Ralith commented:
> it may be a good idea for me to prepare another diff for you to reference so that it's easier for you to see what changed.

Nah, no need to go that far.

patowen commented:
> Nah, no need to go that far.

Got it. I won't go that far, especially if you won't use it. The real diff history will be available in https://github.com/patowen/hypermine/tree/collision-wip, since that's the branch I usually work on, but I won't spend any effort to clean that up.

patowen commented:
> That sounds like a nice improvement. I don't think it needs to block this PR, but if you get it into a state you're happy with before we merge this, then you might as well fold them together.

@Ralith, I ended up making the changes I proposed. It does make some code more verbose and repetitive (especially `chunk_collision::chunk_sphere_cast`), but I would say it's an overall win for readability.

Ralith commented on common/src/node.rs:
```diff
@@ -93,6 +93,52 @@ impl VoxelData {
+
+    /// Converts a single coordinate from dual coordinates in the Klein-Beltrami model to an integer coordinate
+    /// suitable for voxel lookup. Margins are included. Returns `None` if the coordinate is outside the chunk.
+    #[inline]
```
If we're gunna bother to mark things inline, definitely do so for the pure getters above.

Ralith commented on common/src/character_controller.rs:
```diff
@@ -107,7 +107,14 @@ impl CharacterControllerPass<'_> {
+        let cast_hit = graph_collision::sphere_cast(
+            self.cfg.character_radius,
+            self.graph,
+            &ChunkLayout::new(self.cfg.chunk_size as usize),
```
Should we just store the `ChunkLayout` in the config structure directly?

patowen replied:
Possibly. The idea of the `ChunkLayout` struct is that it would eagerly pre-compute things that depend on the dimension, just like `dodeca` pre-computes things that don't depend on the dimension. Using it more widely would have the additional benefit of allowing experimentation with nonlinear layouts where cube-corner voxels are made a bit smaller, and cube-center voxels are made a bit larger.

This design choice has far-reaching implications, and I'm not sure whether it's within scope for this PR. I had decided to put the seam between the current approach and this new `ChunkLayout` approach here, but if you're a fan of this idea, I can add it to `sim_config` in place of `chunk_size`.

Ralith replied:
Let's err on the side of avoiding scope creep.

Ralith commented on common/src/graph_collision.rs:
```diff
@@ -12,39 +16,121 @@ use crate::{
+
+        // Check collision within a single chunk
+        let current_tanh_distance = hit.as_ref().map_or(tanh_distance, |hit| hit.tanh_distance);
+        hit = chunk_sphere_cast(
+            collider_radius,
+            voxel_data,
+            layout,
+            &local_ray,
+            current_tanh_distance,
+        )
+        .map_or(hit, |hit| {
+            Some(GraphCastHit {
+                tanh_distance: hit.tanh_distance,
+                chunk,
+                normal: math::mtranspose(&node_transform)
+                    * chunk.vertex.dual_to_node().cast()
+                    * hit.normal,
+            })
+        });
```
Shouldn't this only replace `hit` if the newly detected hit is closer? Consider e.g. a spherecast roughly parallel to and overlapping wit the edge of a chunk. This would be a good test case, too.

patowen replied:
That is handled by the fact that `chunk_sphere_cast` only detects collisions at `current_tanh_distance` or closer. I agree with adding a test case for this, though.

Ralith replied:
Oh, that's clever, and obvious in retrospect. Didn't read closely enough I guess!

patowen replied:
I went ahead and added the test case.

Ralith commented on common/src/graph_collision.rs:
```diff
@@ -12,39 +16,121 @@ use crate::{
+                let next_node_transform = side.reflection().cast::<f32>() * node_transform;
+                // Crude check to ensure that the neighboring chunk's node can be in the path of the ray. For simplicity, this
+                // check treats each node as a sphere and assumes the ray is pointed directly towards its center. The check is
+                // needed because chunk generation uses this approximation, and this check is not guaranteed to pass near corners.
```
Does this mean `BOUNDING_SPHERE_RADIUS` is incorrect? It sounds like this comment is describing a false negative case, which seems bad.

patowen replied:
`BOUNDING_SPHERE_RADIUS` is correct as far as I can tell. The issue is that finding the exact set of chunks subject to a sphere cast is needlessly complicated, so we use an AABB-based conservative approximation, but this approximation can include chunks outside the set of generated chunks under specific scenarios. This additional check fine-tunes the set of chunks based on a sphere-based conservative approximation instead of an AABB-based one.

Ralith replied:
Ah, I see. Do I understand correctly that we want both because the AABB-based approximation can be much less brute force than doing the sphere-based approximation for every neighbor?

patowen replied:
Yeah. The sphere-based approximation alone is quite brute-force, as it doesn't depend on the direction of the ray at all. It checks whether the bounding sphere and the ray start intersects the bounding sphere of the node, and it doesn't distinguish between chunks within a given node.

The AABB approximation is nice because it checks an adjacent chunk whenever a ray gets too close to the corresponding face of the current chunk, which is quite intuitive and not very expensive to calculate.

patowen replied:
I adjusted the comments a bit to hopefully make this a bit less confusing for future readers of this code.

Ralith commented on common/src/graph_collision.rs:
```diff
@@ -12,39 +16,121 @@ use crate::{
+                // needed because chunk generation uses this approximation, and this check is not guaranteed to pass near corners.
+                let ray_node_distance = (next_node_transform * ray.position).w.acosh();
+                let ray_length = current_tanh_distance.atanh();
+                if ray_node_distance - ray_length
```
Should we also account for the collider radius here?

patowen replied:
That is handled by `> dodeca::BOUNDING_SPHERE_RADIUS as f32 + collider_radius`. However, I should take another look at which terms are on which side of the inequality sign for minimal confusion.

Ralith replied:
Whoops, not sure how I missed that. LGTM.

patowen replied:
There isn't a clear reason why `collider_radius` is on the right side instead of the left, so it is pretty well hidden because of that. I think it's a holdover from when I visualized this as ray tracing with inflated voxels instead of sphere casting with regular voxels.

I've moved it over to the left side where it's easier to see.

Ralith reviewed (changes requested):
Good progress; loved the refactor!

Ralith commented on common/src/graph_collision.rs:
```diff
@@ -0,0 +1,559 @@
+///
+/// This function may return a `SphereCastError` if not enough chunks are generated, even if the ray never reaches an
+/// ungenerated chunk. To prevent these errors, make sure that the distance between the ray's start point and the closest
+/// ungenerated chunk's center is less than `cast_distance + collider_radius + dodeca::BOUNDING_SPHERE_RADIUS`
```
Should this say "greater than", and with respect to the node center, rather than chunk center?

patowen replied:
Good catch. I was picturing as a ray needing to be short enough, hence the "less than", but I did get it backwards.

patowen commented:
Ah, I just today realized that Github has a "Compare" button that works with force pushes. No wonder you didn't need me to prepare a diff.

patowen commented:
Thanks for merging! I'm optimistic that we're past the hardest part of getting player physics that works similarly to the collision demo.

Besides code style, the biggest remaining open questions are how to make sure prediction logic feels right, especially when jumping, but assuming these questions can get resolved, we should be pretty close to being able to close https://github.com/Ralith/hypermine/issues/127. Depending on what criteria we're looking for, we might be already be able to close https://github.com/Ralith/hypermine/issues/34.
